packages/convex-helpers/server/retries.ts

```diff
@@ -0,0 +1,171 @@
+ *
+ * // in a mutation or action
+ * await runWithRetries(ctx, internal.myModule.myAction, { arg1: 123 });
+ * ```
+ */
+import {
+  FunctionReference,
+  FunctionVisibility,
+  Scheduler,
+  FunctionArgs,
+  OptionalRestArgs,
+  getFunctionName,
+  makeFunctionReference,
+  DefaultFunctionArgs,
+  internalMutationGeneric,
+} from "convex/server";
+import { v, ObjectType } from "convex/values";
+
+const DEFAULT_WAIT_BACKOFF = 10;
+const DEFAULT_RETRY_BACKOFF = 10;
+const DEFAULT_BASE = 2;
+const DEFAULT_MAX_FAILURES = 16;
+const retryArguments = {
+  job: v.id("_scheduled_functions"),
+  action: v.string(),
+  actionArgs: v.any(),
+  waitBackoff: v.number(),
+  retryBackoff: v.number(),
+  base: v.number(),
+  maxFailures: v.number(),
+};
+
+/**
+ * Create a function that retries an action with exponential backoff.
+ * e.g.
+ * ```ts
+ * // in convex/utils.ts
+ * import { makeActionRetrier } from "convex-helpers/server/retries";
+ * import { internalMutation } from "./convex/_generated/server";
+ *
+ * export const { runWithRetries, retry } = makeActionRetrier(internalMutation, internal.utils.retry);
+ *
+ * // in a mutation or action
+ * await runWithRetries(ctx, internal.myModule.myAction, { arg1: 123 });
+ * ```
+ *
+ * @param internalMutation From "./convex/_generated/server" or customMutation.
+ * @param retryRef The function reference to the retryRef function exported.
+ * e.g. internal.mymodule.retry
+ * @returns An object with runWithRetries and retry functions.
+ */
+export function makeActionRetrier(retryFnName: string) {
+  const retryRef = makeFunctionReference<
+    "action",
+    ObjectType<typeof retryArguments>
+  >(retryFnName);
+  /**
+   * Run and retry action until it succeeds or fails too many times.
+   *
+   * @param action - Name of the action to run, e.g., `usercode:maybeAction`.
+   * @param actionArgs - Arguments to pass to the action, e.g., `{"failureRate": 0.75}`.
+   * @param options
+   * @param options.waitBackoff=DEFAULT_WAIT_BACKOFF (10) - Initial delay before checking action status, in milliseconds.
+   * @param options.retryBackoff=DEFAULT_RETRY_BACKOFF (10) - Initial delay before retrying, in milliseconds.
+   * @param options.base=DEFAULT_BASE (2) - Base of the exponential backoff.
+   * @param options.maxFailures=DEFAULT_MAX_FAILURES (16) - The maximum number of times to retry the action.
+   */
+  async function runWithRetries<
+    Action extends FunctionReference<
+      "action",
+      Visibility,
+      DefaultFunctionArgs,
+      null | Promise<null> | void | Promise<void>
+    >,
+    Visibility extends FunctionVisibility = "internal"
+  >(
+    ctx: { scheduler: Scheduler },
+    action: Action,
+    actionArgs: FunctionArgs<Action>,
+    options?: {
+      waitBackoff?: number;
+      retryBackoff?: number;
+      base?: number;
+      maxFailures?: number;
+    }
+  ) {
+    const job = await ctx.scheduler.runAfter(
+      0,
+      action,
+      ...([actionArgs] as OptionalRestArgs<Action>)
```

Comment:
this was some annoying typing - it wouldn't let me just pass them in since it wanted sometimes an EmptyObject. I think this can be improved but haven't dug in